### ISSUE TITLE
[feature]: User Model 에 대해서

### ISSUE THREAD

**Issue by feldblume5263** [Chilli-Saewoo/Roll-in#4]
### Feature 설명

User Model을 생성하려고 합니다.

해당 Model은 아마도 Firebase DB에 올라가는 단 하나의 Entity가 될거에요!
(일단 러프하게 제가 생각한거라, 월요일에 VIVI와 멘토링 후 바뀔 수도 있습니다... )

@yeahaluu 
@sy5072 
@lee02029 
@TaekH 

먼저 설명을 좀 드릴게요!
User 은 UUID(user id)와 nickname(String), Stickers ([Sticker]) 을 가집니다.

이런 모양이 될거에요 (Sticker라는 이름은 바뀔 수도 있습니다.)

```
struct User: Equatable {
    let id = UUID()
    let nickname: String
    var stickers: [Sticker] = []
}
```

UUID는 nickname을 유저에게 입력받을 때, 프론트에서 생성해서 하나의 User를 구성합니다.   
이 구성된 User는 파이어베이스에 저장됩니다.   
   
UUID는 어디에 쓰이나면,   
1. **QR코드를 만드는 것에 쓰입니다.** 
QR코드의 주소는 다음과 같습니다.
https://chilli-saewoo.github.io/rollin.github.io/write?[UUID] <- 이런식으로 들어가서 `누구에게 쓰는 스티커인지를 정합니다.`
   
2. **유저를 특정하는 것에 쓰입니다.** <- 이 부분이 핵심
앞서 만들어진 QR코드의 링크로 들어가서 유저에게 Sticker를 작성하면,   
`Sticker가 해당 파이어베이스에서 해당 UUID를 가진 유저의 stickers 배열에 들어가게 됩니다.` 

3. **TF앱에서 본인 ID에 해당하는 정보를 업데이트 합니다.**
앱에서 본인 UUID를 저장해서, 본인 UUID와 일치하는 User의 Stickers 배열이 바뀌는 것을 추적해서 추가될 때마다 앱의 롤링페이퍼를 업데이트 합니다.

.
.
.

파이어베이스에는 User 배열이 들어가게 될텐데   
해당 배열은 아이패드에서도 그대로 쓰입니다.

User 하나 하나마다 `https://chilli-saewoo.github.io/rollin.github.io/write?[UUID]` 이 링크로 QR코드를 생성하고, 밑에 nickname을 붙여줄 거에요! 

.
.
.

일단 그려본 도식은 다음과 같습니다.
참고하셔서 Firebase와 프론트 구성해주시면 될 것 같습니다.

![image](https://user-images.githubusercontent.com/53016167/197328472-e1f746f9-89e3-4237-8b04-4845996b14e2.png)

### Related View(Optional)

모든 View에 영향을 미칩니다.

### 예상 작업 기간

구현이 되어있는 상태입니다.

**Comment by TaekH** [comment]
모델의 Sticker는
- nickname(String)
- message(String)
- sticker(image)
이렇게 이루어지는 거겠죠?

**Comment by feldblume5263** [comment]
> 모델의 Sticker는
> 
> * writerNickname(String)
> * message(String)
> * sticker(image)
>   이렇게 이루어지는 거겠죠?

네 일단은 그렇게 예상하고 있지만, 하루님께서 웹 페이지 만드신 걸 기준으로 iOS에서는 그대로 가져올 생각입니다!

**Comment by yeahaluu** [comment]
깔끔한 설명 감사합니다! 확실히 이해되었습니다.😊👍
의문이 드는 부분은 Stiker를 다른 Entity로 분리해야하는 것이 나을것 같아 고민됩니다. Stiker가 늘어남에 따라 user id와 nickname도 불필요하게 중복이 될 것 같아서요. 이 부분에 대해서 어떻게 생각하실까요?!

**Comment by feldblume5263** [comment]
> 깔끔한 설명 감사합니다! 확실히 이해되었습니다.😊👍 의문이 드는 부분은 Stiker를 다른 Entity로 분리해야하는 것이 나을것 같아 고민됩니다. Stiker가 늘어남에 따라 user id와 nickname도 불필요하게 중복이 될 것 같아서요. 이 부분에 대해서 어떻게 생각하시나요?!

Sticker가 늘어나면 user id와 nickname이 중복된다는 것이 잘 이해가 가질 않는데, Sticker가 user id와 nickname 정보를 가지고 있지 않기 때문에 중복이 안될것 같다고 생각했거든요!
제가 생각하기에는 분리를 하면, 오히려 sticker 내부에 매칭하는 user id를 가지고 있게 될 것 같은데 조금 더 설명해주실 수 있으실까요?? 🥹

**Comment by yeahaluu** [comment]
아! 저는 웹에서 Stiker가 추가할 때마다, Entity에 하나의 열이 더 추가된다고 생각했는데, Stickers안에 여러 개의 Sticker의 정보가 들어가는 것이군요! 그러면 user id와 nickname이 중복되지는 않을 것 같습니다.

**Comment by yeahaluu** [comment]
아! 저는 웹에서 Stiker가 추가할 때마다, Entity에 하나의 열이 더 추가된다고 생각했는데, Stickers안에 여러 개의 Sticker의 정보가 들어가는 것이군요! 그러면 user id와 nickname이 중복되지는 않을 것 같습니다. 추가 설명으로 이해되었습니다 감사합니다😊
제가 말씀 드린 부분에 대해서 설명드리자면, 'stikers entity 내부에 매칭하는 user id'를 fk(foreign key)로 사용할 수 있다고 생각했습니다. 물론 fk를 사용하는 것이 데이터무결성 측면에선 좋지만, 성능 면에서 좋지 않을 수 있습니다. 하지만 fk를 쓰는 것도 괜찮은 방법이라고 생각합니다!
결론적으로 스닥이 구성한 모델이 이해되었고 이렇게 해보는 것은 좋은 것 같습니다. 하지만 아직 Stikers의 배열이 효과적일지 대해서는 아직 가시적이지 않아서 감이 오지 않고, 그 이유가 Firebase와 NoSQL, DB와 fk에 대한 지식이 부족해서인 것 같습니다. 더 공부를 한 후 더 궁금한 것이 있다면 코멘트 남기겠습니다! 그리고 비비와도 멘토링때 저도 참석하고 싶습니다🙌

**Comment by feldblume5263** [comment]
> 아! 저는 웹에서 Stiker가 추가할 때마다, Entity에 하나의 열이 더 추가된다고 생각했는데, Stickers안에 여러 개의 Sticker의 정보가 들어가는 것이군요! 그러면 user id와 nickname이 중복되지는 않을 것 같습니다. 추가 설명으로 이해되었습니다 감사합니다😊 제가 말씀 드린 부분에 대해서 설명드리자면, 'stikers entity 내부에 매칭하는 user id'를 fk(foreign key)로 사용할 수 있다고 생각했습니다. 물론 fk를 사용하는 것이 데이터무결성 측면에선 좋지만, 성능 면에서 좋지 않을 수 있습니다 하지만 fk를 쓰는 것도 괜찮은 방법이라고 생각합니다! 결론적으로 스닥이 구성한 모델이 이해되었고 이렇게 해보는 것은 좋은 것 같습니다. 하지만 아직 Stikers의 배열이 효과적일지 대해서는 아직 가시적이지 않아서 감이 오지 않고, 그 이유가 Firebase와 NoSQL, DB와 fk에 대한 지식이 부족해서인 것 같습니다. 더 공부를 한 후 더 궁금한 것이 있다면 코멘트 남기겠습니다! 그리고 비비와도 멘토링때 저도 참석하고 싶습니다🙌

NoSQL에서 가장 효과적인 방법을 찾아서 하는게 매우 중요한 것 같습니다! 쉐리님께서 FK를 제시해주셔서 저도 찾아보았는데 확실히 무결성 측면에서 좋을 것 같고, 나중에 기능이 확장되었을 때 스티커 자체에서도 정보를 가지고 있다는 면에서 확장성도 생길 것 같은데 맞을까요?😅 
프론트적인 시작에서 배열을 사용한건데, 조금 검색을 하다보니 Entity를 따로 관리했을 때 얻을 수 있는 이점이 많은 것 같아서 꼭 이 부분에 대한 고민이 필요하다는 생각이 들었습니다.
넵 비비와의 멘토링도 꼭 함께해요!!ㅎㅎ

**Comment by yeahaluu** [comment]
Firebase의 Realtime DB의 경우 table이 아닌 json트리로 데이터가 구조화되어 있습니다. 따라서 스닥이 User Model에 대해 고민해주신것을 반영하여 아래와 같은 구조로 json 데이터를 쓰고(웹, 앱) 읽는 것(앱)을 제안드립니다.
```
{
  "users": {
    "UUID1(xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx)": {
      "nickname": "sherry",
        "stickers": {
	  "s1": { "sender": "Noasdaq", "image": "1", "message": "안녕" },
	  "s2": { "sender": "Taek", "image": "4", "message": "메시지" },
	},
    },
    "UUID2(xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx)": { ... },
  }
}
```
데이터 쓰고 읽는 방식은 다음 공식문서에서 확인할 수 있습니다.
앱: https://firebase.google.com/docs/database/ios/read-and-write?hl=ko
웹: https://firebase.google.com/docs/database/web/read-and-write?hl=ko
